### PR TITLE
Simplify HivePageSink

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java [hunk 1]
```diff
@@ -675,53 +714,46 @@ private static Page extractColumns(Page page, int[] columns)
...
+            structFields = ImmutableList.copyOf(inputColumnNames.stream()
+                    .map(tableInspector::getStructFieldRef)
+                    .collect(toList()));
 
             row = tableInspector.create();
 
             setters = new FieldSetter[structFields.size()];
             for (int i = 0; i < setters.length; i++) {
-                setters[i] = createFieldSetter(tableInspector, row, structFields.get(i), inputColumns.get(i).getType());
+                setters[i] = createFieldSetter(tableInspector, row, structFields.get(i), fileColumnTypes.get(structFields.get(i).getFieldID()));
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java]: Can we validate that the corresponding inputColumn.type match up?

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java [hunk 2]
```diff
@@ -242,52 +187,42 @@ private void doAbort()
...
-                if (writer == null) {
-                    if (bucketWriterCount >= maxOpenPartitions) {
-                        throw new PrestoException(HIVE_TOO_MANY_OPEN_PARTITIONS, "Too many open writers for partitions and buckets");
-                    }
-                    bucketWriterCount++;
-                    writer = writerFactory.createWriter(partitionColumns, position, OptionalInt.of(bucket));
-                    writers.put(bucket, writer);
-                }
+        // create missing writers
+        while (writers.size() <= pagePartitioner.getMaxIndex()) {
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java]: I would replace this line and the next with a for loop.

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java [hunk 3]
```diff
@@ -242,52 +187,42 @@ private void doAbort()
...
+            int newWriterIndex = writers.size();
+            int position = Ints.indexOf(writerIndexes, newWriterIndex);
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java]: `page` can potentially have many rows. Let's not write O(N^2) code. (N is the number of rows. Every row is a unique partition.) Given the low maxOpenWriters right now, it probably won't cause a problem. But this feels like a land mine.

I would do:
```
  for (int position = 0; position < page.getPositionCount(); position++) {
    if (writers.size() > pagePartitioner.getMaxIndex()) {
      break;
    }
    if (writerIndexes[i] < writers.size()) {
      continue;
    }
    verify(position == writers.size());
    // create the writer
    writers.add(writer);
  }
  verify(writers.size() == pagePartitioner.getMaxIndex() + 1;
```

Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java]: I'm a bit concerned that this requires that the page partitioner adds new indexes as it progresses thought the pages.  What do you think about this version:
```java
        // expand writers list to new size
        while (writers.size() <= pagePartitioner.getMaxIndex()) {
            writers.add(null);
        }

        // create missing writers
        for (int position = 0; position < page.getPositionCount(); position++) {
            int writerIndex = writerIndexes[position];
            if (writers.get(writerIndex) != null) {
                continue;
            }

            OptionalInt bucketNumber = OptionalInt.empty();
            if (bucketBlock != null) {
                bucketNumber = OptionalInt.of(bucketBlock.getInt(position, 0));
            }
            HiveWriter writer = writerFactory.createWriter(partitionColumns, position, bucketNumber);

            writers.set(writerIndex, writer);
        }

        verify(writers.size() == pagePartitioner.getMaxIndex() + 1);
        verify(!writers.contains(null));
```

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java (hunk 1 above)
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java]: I'm not sure what you mean by `inputColumn.type`

**Pull request closed without merging**